bpf/rtt_tracker.h

```diff
@@ -1,143 +1,161 @@
+    u8 *skb_head = BPF_CORE_READ(skb, head);
+
+    switch (family) {
+    case AF_INET: {
+        struct iphdr ip;
+        __builtin_memset(&ip, 0, sizeof(ip));
+        bpf_probe_read(&ip, sizeof(ip), (struct iphdr *)(skb_head + skb_network_header));
+        __builtin_memcpy(id->src_ip, ip4in6, sizeof(ip4in6));
+        __builtin_memcpy(id->dst_ip, ip4in6, sizeof(ip4in6));
+        __builtin_memcpy(id->src_ip + sizeof(ip4in6), &ip.saddr, sizeof(ip.saddr));
+        __builtin_memcpy(id->dst_ip + sizeof(ip4in6), &ip.daddr, sizeof(ip.daddr));
+        *dscp = ipv4_get_dscp(&ip);
+        break;
+        }
+    case AF_INET6: {
+        struct ipv6hdr ip;
+        __builtin_memset(&ip, 0, sizeof(ip));
+        bpf_probe_read(&ip, sizeof(ip), (struct ipv6hdr *)(skb_head + skb_network_header));
+        __builtin_memcpy(id->src_ip, ip.saddr.in6_u.u6_addr8, IP_MAX_LEN);
+        __builtin_memcpy(id->dst_ip, ip.daddr.in6_u.u6_addr8, IP_MAX_LEN);
+        *dscp = ipv6_get_dscp(&ip);
+        break;
         }
+    default:
+        return;
     }
 }
 
-static __always_inline void __calculate_tcp_rtt(pkt_info *pkt, struct tcphdr *tcp, flow_seq_id *seq_id) {
-    // Stored sequence should be ack_seq - 1
-    u32 seq = bpf_ntohl(tcp->ack_seq) - 1;
-    // check reversed flow
-    fill_flow_seq_id(seq_id, pkt, seq, true);
-
-    u64 *prev_ts = (u64 *)bpf_map_lookup_elem(&flow_sequences, seq_id);
-    if (prev_ts != NULL) {
-        u64 rtt = pkt->current_ts - *prev_ts;
-        /**
-         * FIXME: Because of SAMPLING the way it is done if we miss one of SYN/SYN+ACK/ACK
-         * then we can get RTT values which are the process response time rather than actual RTT.
-         * This check below clears them out but needs to be modified with a better solution or change
-         * the algorithm for calculating RTT so it doesn't interact with SAMPLING like this.
-         */
-        if (rtt < MIN_RTT) {
-            return;
+static inline void rtt_fill_in_tcp(struct sk_buff *skb, flow_id *id, u16 *flags) {
+    u16 skb_transport_header = BPF_CORE_READ(skb, transport_header);
+    u8 *skb_head = BPF_CORE_READ(skb, head);
+    struct tcphdr tcp;
+    u16 sport, dport;
+
+    __builtin_memset(&tcp, 0, sizeof(tcp));
+
+    bpf_probe_read(&tcp, sizeof(tcp), (struct tcphdr *)(skb_head + skb_transport_header));
+    sport = bpf_ntohs(tcp.source);
+    dport = bpf_ntohs(tcp.dest);
+    id->src_port = sport;
+    id->dst_port = dport;
+    set_flags(&tcp, flags);
+    id->transport_protocol = IPPROTO_TCP;
+}
+
+static inline int rtt_lookup_and_update_flow(flow_id *id, u16 flags, u64 rtt) {
+    flow_metrics *aggregate_flow = bpf_map_lookup_elem(&aggregated_flows, id);
+    if (aggregate_flow != NULL) {
+        aggregate_flow->end_mono_time_ts = bpf_ktime_get_ns();
+        aggregate_flow->flags |= flags;
+        if (aggregate_flow->flow_rtt < rtt) {
+            aggregate_flow->flow_rtt = rtt;
         }
```

Comment:
So this means we keep always the highest srtt per flow ... which seems legit to me as we're generally interested in finding high latencies - but that might be something to keep in mind for documentation, if we document that level of details.

Reply:
Yes I will and this is something I am debating with and waiting on rhel folks input other implementations aggregate rtt value I think getting the max is appropriate but I need confirmation and ofc we will reflect that in the doc and rtt blog